exchanges/coinbasepro/coinbasepro_test.go

```diff
@@ -2,679 +2,1488 @@ package coinbasepro
-		t.Errorf("GetFee() error. Expected: %f, Received: %f", float64(0.002), resp)
-	}
+func TestSendAuthenticatedHTTPRequest(t *testing.T) {
+	t.Parallel()
+	fc := &CoinbasePro{}
+	err := fc.SendAuthenticatedHTTPRequest(context.Background(), exchange.EdgeCase3, "", "", nil, nil, false, nil, nil)
+	assert.ErrorIs(t, err, exchange.ErrCredentialsAreEmpty)
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, c)
+	err = c.SendAuthenticatedHTTPRequest(context.Background(), exchange.EdgeCase3, "", "", nil, nil, false, nil, nil)
+	if err.Error() != errNoEndpointPathEdgeCase3 {
+		t.Errorf(errExpectMismatch, err, errNoEndpointPathEdgeCase3)
+	}
+	ch := make(chan struct{})
+	body := map[string]interface{}{"Unmarshalable": ch}
+	err = c.SendAuthenticatedHTTPRequest(context.Background(), exchange.RestSpot, "", "", nil, body, false, nil, nil)
+	var targetErr *json.UnsupportedTypeError
+	assert.ErrorAs(t, err, &targetErr)
+}
 
-	// high volume
-	volume = []Volume{
-		{
-			ProductID: "btc_USD",
-			Volume:    100000010000,
-		},
+func TestGetFee(t *testing.T) {
+	t.Parallel()
+	_, err := c.GetFee(context.Background(), nil)
+	assert.ErrorIs(t, err, common.ErrNilPointer)
+	feeBuilder := exchange.FeeBuilder{
+		FeeType:       exchange.OfflineTradeFee,
+		Amount:        1,
+		PurchasePrice: 1,
 	}
-
-	if resp := c.calculateTradingFee(volume, currency.BTC, currency.USD, "_", 1, 1, false); resp != float64(0.001) {
-		t.Errorf("GetFee() error. Expected: %f, Received: %f", float64(0.001), resp)
+	resp, err := c.GetFee(context.Background(), &feeBuilder)
+	assert.NoError(t, err)
+	if resp != WorstCaseTakerFee {
+		t.Errorf(errExpectMismatch, resp, WorstCaseTakerFee)
+	}
+	feeBuilder.IsMaker = true
+	resp, err = c.GetFee(context.Background(), &feeBuilder)
+	assert.NoError(t, err)
+	if resp != WorstCaseMakerFee {
+		t.Errorf(errExpectMismatch, resp, WorstCaseMakerFee)
+	}
+	feeBuilder.Pair = currency.NewPair(currency.USDT, currency.USD)
+	resp, err = c.GetFee(context.Background(), &feeBuilder)
+	assert.NoError(t, err)
+	if resp != 0 {
+		t.Errorf(errExpectMismatch, resp, StablePairMakerFee)
+	}
+	feeBuilder.IsMaker = false
+	resp, err = c.GetFee(context.Background(), &feeBuilder)
+	assert.NoError(t, err)
+	if resp != WorstCaseStablePairTakerFee {
+		t.Errorf(errExpectMismatch, resp, WorstCaseStablePairTakerFee)
 	}
-
-	// no match
-	volume = []Volume{
-		{
-			ProductID: "btc_beeteesee",
-			Volume:    100000010000,
-		},
+	feeBuilder.FeeType = exchange.CryptocurrencyDepositFee
+	_, err = c.GetFee(context.Background(), &feeBuilder)
+	assert.ErrorIs(t, err, errFeeTypeNotSupported)
+	feeBuilder.Pair = currency.Pair{}
+	sharedtestvalues.SkipTestIfCredentialsUnset(t, c)
+	feeBuilder.FeeType = exchange.CryptocurrencyTradeFee
+	resp, err = c.GetFee(context.Background(), &feeBuilder)
+	assert.NoError(t, err)
+	if !(resp <= WorstCaseTakerFee && resp >= BestCaseTakerFee) {
+		t.Errorf(errExpectedFeeRange, BestCaseTakerFee, WorstCaseTakerFee, resp)
```

Comment:
I _think_ that's important since the worst case fees are returned by GetFee() sometimes, and we'd want to know when those are giving incorrect answers.

If you think that doesn't justify having stuff like this error out sometimes, let me know, and I'll change the test as you suggest.

Reply:
Just requested for you. Do you have API keys?

```
{"total_volume":0, "total_fees":0, "fee_tier":{"pricing_tier":"Advanced 1", "usd_from":"0", "usd_to":"1000", "taker_fee_rate":"0.01", "maker_fee_rate":"0.006", "aop_from":"", "aop_to":""}, "margin_rate":null, "goods_and_services_tax":{"rate":"0.1", "type":"EXCLUSIVE"}, "advanced_trade_only_volume":0, "advanced_trade_only_fees":0, "coinbase_pro_volume":0, "coinbase_pro_fees":0, "total_balance":"", "has_promo_fee":false}
```

`coinbasepro_test.go:1063: expected fee range of 0.0005 and 0.008, received 0.01`

The same occurs even [when using the non-deprecated fields](https://docs.cloud.coinbase.com/advanced-trade/reference/retailbrokerageapi_gettransactionsummary):
`https://api.coinbase.com/api/v3/brokerage/transaction_summary?product_type=SPOT`

```
{"total_volume":0, "total_fees":0, "fee_tier":{"pricing_tier":"Advanced 1", "usd_from":"0", "usd_to":"1000", "taker_fee_rate":"0.01", "maker_fee_rate":"0.006", "aop_from":"", "aop_to":""}, "margin_rate":null, "goods_and_services_tax":{"rate":"0.1", "type":"EXCLUSIVE"}, "advanced_trade_only_volume":0, "advanced_trade_only_fees":0, "coinbase_pro_volume":0, "coinbase_pro_fees":0, "total_balance":"", "has_promo_fee":false}
 ```

Documentation on an exchange gets out of date quickly. API endpoints can change without documentation updates. 
It's why I don't recommend a test rely on specific numbers like above. It is not the fault of us that something has changed on the exchange's end.

Reply:
Fees are inline with my account page:
<img width="1703" alt="image" src="https://github.com/thrasher-corp/gocryptotrader/assets/9261323/f59fede6-b0e1-4974-9158-16ed91d679a4">
1% taker = 0.01

Reply:
Yeah, got API keys but am getting different fee amounts from the same endpoints for the same trading amounts. Very strange.

Anyway, I take it you just want me to do the >0 stuff; that looks like it'll work for taker but should I still run that with maker, or do something like >-0.01?